### PR TITLE
display layers meta data (layer info)

### PR DESCRIPTION
**Pull request by thevoiceofthruth**
new components for loading layerinfo

### CODE REVIEW

**Review by taulinger — COMMENTED** [review]
Looks good so far 👍
Let's discuss tomorrow, how to proceed...

File: src/modules/layerInfo/components/LayerInfoPanel.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+	/**
+	 * @override
+	 */
+	createView(model) {
+		const { layerInfo } = model;
+
+		if (layerInfo) {
+			return html`
+			<div>${unsafeHTML(`${layerInfo.content}`)} ${layerInfo.title}</div>
```
Comment (taulinger) [inline src/modules/layerInfo/components/LayerInfoPanel.js:36]: Let's render the title in its own div tag, if present

File: test/modules/layerManager/components/LayerItem.test.js [hunk 2]
```diff
@@ -179,6 +180,18 @@ describe('LayerItem', () => {
...
+		it('click on info icon show layerinfo panel as modal', async () => {
+			const store = setup();
+			const element = await TestUtils.render(LayerItem.tag);
+			element.layer = { ...layer };
+
+			const infoButton = element.shadowRoot.querySelector('#info');
+
+			infoButton.click();
+
+			expect(store.getState().modal.data.title).toBe('label0');
```
Comment (taulinger) [inline test/modules/layerManager/components/LayerItem.test.js:193]: test for `modal.data.content` is missing

File: test/modules/topics/components/menu/catalog/CatalogLeaf.test.js [hunk 3]
```diff
@@ -157,6 +158,24 @@ describe('CatalogLeaf', () => {
...
+					//load leaf data
+					const leaf = (await loadExampleCatalog('foo')).pop();
+					const element = await setup('foo', []);
+					//assign data
+					element.data = leaf;
+					const icon = element.shadowRoot.querySelector('ba-icon');
+
+					icon.click();
+
+					expect(store.getState().modal.data.title).toBe('someLabel');
```
Comment (taulinger) [inline test/modules/topics/components/menu/catalog/CatalogLeaf.test.js:176]: test for modal.data.content is missing

**Review by taulinger — COMMENTED** [review]
Nice proceedings, added some inline comments...

Next week we will extend the `LayerItem` to hold a `geoResourceId` and 
and make some thoughts about improving the UX while loading data from the backend.

File: src/services/provider/layerInfo.provider.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { $injector } from '../../injection';
+
+/**
+ * Uses the BVV endpoint to load layerinfo.
+ * @function
+ * @returns {Promise<Object>}
+ */
+export const loadBvvLayerInfo = async (geoResourceId) => {
```
Comment (taulinger) [inline src/services/provider/layerInfo.provider.js:8]: Let's return a LayerInfoResult instance also in our provider

File: src/modules/layerInfo/services/layerInfo.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+/**
+* @class
+* @author costa_gi
+*/
+export class LayerInfo {
```
Comment (taulinger) [inline src/modules/layerInfo/services/layerInfo.js:5]: Let's name the class LayerInfoResult (we want to name all of our backend response data classes "XYResult" in the future)

**Review by taulinger — COMMENTED** [review]
see inline comments

File: src/modules/layerInfo/services/layerInfo.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	}
+
+	get content() {
+		return this._content;
+	}
+
+	get title() {
+		return this._title;
+	}
+}
```
[the remark this answers is not in the document]
Comment (thevoiceofthruth) [reply src/modules/layerInfo/services/layerInfo.js:25]: do you mean the conntent of layerInfo.js has to be inserted in LayerInfoService.js?

[the remark this answers is not in the document]
Comment (taulinger) [reply src/modules/layerInfo/services/layerInfo.js:25]: Yes, exactly

File: test/modules/layerInfo/components/LayerInfoPanel.test.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+			element.geoResourceId = '914c9263-5312-453e-b3eb-5104db1bf788';
+			element.signal('UPDATE_LAYERINFO', layerInfo);
+			const divs = element.shadowRoot.querySelectorAll('div');
+
+			expect(divs.length).toBe(2);
+			expect(divs[1].innerText).toBe('content');
+		});
+
+		it('should log a warn statement when Altitude Service is not available', async (done) => {
```
Comment (taulinger) [inline test/modules/layerInfo/components/LayerInfoPanel.test.js:41]: must be LayerInfoService here

File: test/modules/layerInfo/components/LayerInfoPanel.test.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+			spyOn(layerInfoServiceMock, 'byId').withArgs('914c9263-5312-453e-b3eb-5104db1bf788').and.returnValue(Promise.reject(new Error('LayerInfo Error')));
+			const warnSpy = spyOn(console, 'warn');
+
+			const element = await TestUtils.render(LayerInfoPanel.tag);
+			element.geoResourceId = '914c9263-5312-453e-b3eb-5104db1bf788';
+			element.signal('UPDATE_LAYERINFO', null);
```
Comment (taulinger) [inline test/modules/layerInfo/components/LayerInfoPanel.test.js:48]: no need to call `signal()` here

File: test/modules/layerInfo/components/LayerInfoPanel.test.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+		it('should show a layerInfo on the panel', async () => {
+
+			const layerInfo = new LayerInfoResult('<b>content</b>');
+			spyOn(layerInfoServiceMock, 'byId').withArgs('914c9263-5312-453e-b3eb-5104db1bf788').and.returnValue(layerInfo);
+
+			const element = await TestUtils.render(LayerInfoPanel.tag);
+
+			element.geoResourceId = '914c9263-5312-453e-b3eb-5104db1bf788';
+			element.signal('UPDATE_LAYERINFO', layerInfo);
```
Comment (taulinger) [inline test/modules/layerInfo/components/LayerInfoPanel.test.js:34]: no need to call signal() here

**Review by taulinger — COMMENTED** [review]
Have a look at all yet unresolved conversions

**Review by taulinger — COMMENTED** [review]
Thx!
Please have a look at the remaining unresolved conversations, e.g.
https://github.com/ldbv-by/bav4-nomigration/pull/519#discussion_r743815653

**Comment by taulinger** [comment]
Tests for modal.data.content in `LayerItem.test` and `CatalogLeaf.test` are good!
Didn't recognize the changes, sorry, my mistake. 
Resolved the corresponding conversations.

File: src/modules/layerManager/components/LayerItem.js [hunk 10]
```diff
@@ -132,7 +132,7 @@ export class LayerItem extends AbstractContentPanel {
 
 		const openLayerInfoPanel = async () => {
 			const content = html`<ba-layerinfo-panel .geoResourceId=${this._layer.id}></ba-layerinfo-panel>`;
-			openModal(this._layer.label, content);
+			openModal(this._layer.geoResourceId, content);
```
Comment (taulinger) [inline src/modules/layerManager/components/LayerItem.js:135]: Here we have to use the label of the GeoResource instance as the title for the modal.
The label we get by:
```
const geoR = this._geoResourceService.byId(geoResourceId);
const label = geoR ? geoR.label : geoResourceId;

```
see `CatalogLeaf.js` line 43

**Review by taulinger — COMMENTED** [review]
Thx, for the next step, have a look at the comment

**Review by taulinger — COMMENTED** [review]
Looks good, thx.

Now we want to give the user a loading hint while the backend response is pending yet.
You can use the `<ba-spinner>` component for this case.

File: src/injection/config.js [hunk 11]
```diff
@@ -50,6 +51,7 @@ $injector
 	.register('UrlService', UrlService)
 	.register('AdministrationService', AdministrationService)
 	.register('FeatureInfoService', FeatureInfoService)
+	.registerSingleton('LayerInfoService', new LayerInfoService())
```
Comment (taulinger) [inline src/injection/config.js:54]: Should have "per Instance scope"

File: test/modules/layerInfo/services/LayerInfoService.test.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+describe('LayerInfoService', () => {
+
+	it('initializes the service with default provider', async () => {
+
+		const layerInfoService = new LayerInfoService();
+
+		expect(layerInfoService._provider).toEqual(loadBvvLayerInfo);
+	});
+
```
Comment (taulinger) [inline test/modules/layerInfo/services/LayerInfoService.test.js:14]: We also could add a test for a custom provider, e.g.
```
	it('initializes the service with custom provider', async () => {
			const customProvider = async () => { };
			const instanceUnderTest = setup(customProvider);
			expect(instanceUnderTest._provider).toBeDefined();
			expect(instanceUnderTest._provider).toEqual(customProvider);
		});
```

File: src/services/provider/layerInfoResult.provider.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+			const htmlContent = await result.text();
+			return new LayerInfoResult(htmlContent);
+		}
+		case 404: {
+			return null;
+		}
+	}
+
+	throw new Error(`LayerInfoResult for '${geoResourceId}' could not be loaded`);
+};
```
Comment (taulinger) [inline src/services/provider/layerInfoResult.provider.js:27]: File must be located in `src/modules/layerInfo/services/provider`.
Same for the corresponding test file.

**Review by taulinger — COMMENTED** [review]
Looks good, have a look at the comments, please.
After that, we will decide about the next steps.

File: src/modules/layerInfo/i18n/layerInfo.provider.js [hunk 14]
```diff
@@ -0,0 +1,21 @@
+export const provide = (lang) => {
+	switch (lang) {
+
+		case 'en':
+			return {
+				//the first part of the snake_case key should be the name of the related module
+				layerinfo_empty_layerInfo: 'No information for the layer found',
+				layerinfo_layerInfo_response_error: 'The layer-information could not be loaded'
```
Comment (taulinger) [inline src/modules/layerInfo/i18n/layerInfo.provider.js:8]: would suggest `The Layer Information could not be loaded`

File: src/modules/layerInfo/i18n/layerInfo.provider.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
+export const provide = (lang) => {
+	switch (lang) {
+
+		case 'en':
+			return {
+				//the first part of the snake_case key should be the name of the related module
+				layerinfo_empty_layerInfo: 'No information for the layer found',
```
Comment (taulinger) [inline src/modules/layerInfo/i18n/layerInfo.provider.js:7]: would suggest `No Layer Information available`

File: src/modules/layerInfo/i18n/layerInfo.provider.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+				layerinfo_layerInfo_response_error: 'The layer-information could not be loaded'
+			};
+
+		case 'de':
+			return {
+				//the first part of the snake_case key should be the name of the related module
+				layerinfo_empty_layerInfo: 'Keine Informationen für die Ebene gefunden',
```
Comment (taulinger) [inline src/modules/layerInfo/i18n/layerInfo.provider.js:14]: `Kein Ebenen-Information verfügbar`

**Review by taulinger — COMMENTED** [review]
added some suggestions for better i18n translations

**Review by taulinger — COMMENTED** [review]
Thx 👍, added one comment.
The next thing will be a fallback LayerInfo mechanism for our fallback Georesources
'atkis', 'atkis_sw'

**Review by taulinger — COMMENTED** [review]
Thx

**Review by taulinger — APPROVED** [review]
👍 Let's merge the branch

**Comment by thevoiceofthruth** [comment]
layerinfo functionality added

**Pull request merged**